src/Control/Effect/Carrier.hs

```diff
@@ -34,46 +31,109 @@ class HFunctor h where
-         -> sig m (m a)
-         -> sig n (n (f a))
-
-  default handle :: (Functor f, Coercible (sig m (n (f a))) (sig n (n (f a))))
+         -> sig m a
+         -> sig n (f a)
+  default handle :: (Functor f, Monad m, Monad n, Generic1 (sig m), Generic1 (sig n), GEffect m n (Rep1 (sig m)) (Rep1 (sig n)))
                  => f ()
                  -> (forall x . f (m x) -> n (f x))
-                 -> sig m (m a)
-                 -> sig n (n (f a))
-  handle state handler = coerce . fmap' (handler . (<$ state))
+                 -> sig m a
+                 -> sig n (f a)
+  handle state handler = to1 . ghandle state handler . from1
   {-# INLINE handle #-}
 
 
-
 -- | The class of carriers (results) for algebras (effect handlers) over signatures (effects), whose actions are given by the 'eff' method.
 class (HFunctor sig, Monad m) => Carrier sig m | m -> sig where
   -- | Construct a value in the carrier for an effect signature (typically a sum of a handled effect and any remaining effects).
-  eff :: sig m (m a) -> m a
+  eff :: sig m a -> m a
 
 -- | Apply a handler specified as a natural transformation to both higher-order and continuation positions within an 'HFunctor'.
-handlePure :: HFunctor sig => (forall x . f x -> g x) -> sig f (f a) -> sig g (g a)
-handlePure handler = hmap handler . fmap' handler
+handlePure :: (HFunctor sig, Functor f, Functor g) => (forall x . f x -> g x) -> sig f a -> sig g a
+handlePure = hmap
 {-# INLINE handlePure #-}
+{-# DEPRECATED handlePure "handlePure has been subsumed by hmap." #-}
 
 -- | Thread a 'Coercible' carrier through an 'HFunctor'.
 --
 --   This is applicable whenever @f@ is 'Coercible' to @g@, e.g. simple @newtype@s.
-handleCoercible :: (HFunctor sig, Coercible f g) => sig f (f a) -> sig g (g a)
-handleCoercible = handlePure coerce
+handleCoercible :: (HFunctor sig, Functor f, Functor g, Coercible f g) => sig f a -> sig g a
+handleCoercible = hmap coerce
 {-# INLINE handleCoercible #-}
+
+
+-- | Generic implementation of 'HFunctor'.
+class GHFunctor m m' rep rep' where
+  -- | Generic implementation of 'hmap'.
+  ghmap :: (Functor m, Functor m') => (forall x . m x -> m' x) -> (rep a -> rep' a)
+
+instance GHFunctor m m' rep rep' => GHFunctor m m' (M1 i c rep) (M1 i c rep') where
+  ghmap f = M1 . ghmap f . unM1
+
+instance (GHFunctor m m' l l', GHFunctor m m' r r') => GHFunctor m m' (l :+: r) (l' :+: r') where
+  ghmap f (L1 l) = L1 (ghmap f l)
+  ghmap f (R1 r) = R1 (ghmap f r)
+
+instance (GHFunctor m m' l l', GHFunctor m m' r r') => GHFunctor m m' (l :*: r) (l' :*: r') where
+  ghmap f (l :*: r) = ghmap f l :*: ghmap f r
+
+instance GHFunctor m m' V1 V1 where
+  ghmap _ v = case v of {}
+
+instance GHFunctor m m' U1 U1 where
+  ghmap _ = id
+
+instance GHFunctor m m' (K1 R c) (K1 R c) where
+  ghmap _ = coerce
+
+instance (Functor f, GHFunctor m m' g g') => GHFunctor m m' (f :.: g) (f :.: g') where
+  ghmap f = Comp1 . fmap (ghmap f) . unComp1
```

Comment:
There are other configurations possible here that this instance doesn’t cover, e.g. `m (f a)`, `m (m a)`, `m (f (m a))`, etc., but I wanted to avoid getting into overlapping instances.

Reply:
I think this is fair. More delicate instances can be covered by `fused-effects-th`, when it’s done.